anvio/synteny.py

```diff
@@ -0,0 +1,303 @@
+import anvio.tables as t
+import anvio.terminal as terminal
+import anvio.filesnpaths as filesnpaths
+import anvio.genomedescriptions as genomedescriptions
+
+from anvio.errors import ConfigError
+
+
+__author__ = "Developers of anvi'o (see AUTHORS.txt)"
+__copyright__ = "Copyleft 2015-2018, the Meren Lab (http://merenlab.org/)"
+__credits__ = []
+__license__ = "GPL 3.0"
+__version__ = anvio.__version__
+__maintainer__ = "Matthew Schechter"
+__email__ = "mschechter@uchicago.edu"
+
+
+class NGram(object):
+    """class for counting NGrams
+
+    anvi-analyze-synteny is designed to work with a group of similar loci, where each locus is a contig (which
+    can lie in any number of contigs dbs)
+
+    Parameters
+    ==========
+    args : argparse.Namespace
+        For examples, arguments accepted by anvi-analyze-syntenty
+
+    skip_sanity_check : bool, False
+        If True, sanity_check will not be called.
+
+    Notes
+    =====
+    - Currently the design assumes that each locus is a contig. In the future we have plans to expand this to
+      compare genomes, not just loci, to each other. If that behavior is desired in the current design, each
+      genome should be a single contig.
+    """
+
+    def __init__(self, args, run=terminal.Run(), progress=terminal.Progress(), skip_sanity_check=False):
+        """Parses arguments and run sanity_check"""
+
+        self.args = args
+        self.run = run
+        self.progress = progress
+
+        # Parse arguments
+        A = lambda x: args.__dict__[x] if x in args.__dict__ else None
+        self.external_genomes = A('external_genomes')
+        self.annotation_source = A('annotation_source')
+        self.window_range = A('window_range')
+        self.in_in_unknowns_mode = A('analyze_unknown_functions')
+        self.external_genomes = utils.get_TAB_delimited_file_as_dictionary(self.external_genomes)
+        self.output_file = A('output_file')
+
+        # This houses the ngrams' data
+        self.ngram_attributes_list = []
+
+        self.num_contigs_in_external_genomes_with_genes = 0
+
+        if not skip_sanity_check:
+            self.sanity_check()
+
+        # unless we are in debug mode, let's keep things quiet.
+        if anvio.DEBUG:
+            self.run_object = terminal.Run()
+        else:
+            self.run_object = terminal.Run(verbose=False)
+
+
+    def sanity_check(self):
+        """Sanity_check will confirm input for NGram class"""
+
+        # checking if the annotation source is common accross all contigs databases
+        g = genomedescriptions.GenomeDescriptions(self.args)
+        g.load_genomes_descriptions(init=False)
+
+        if self.annotation_source not in g.function_annotation_sources:
+            raise ConfigError("The annotation source you requested does not appear to be in the "
+                              "contigs database :/ "
+                              "Please confirm you are only providing one annotation source :)")
```

Comment:
This error message needs to be updated. `g.function_annotation_sources` shows sources for functions that are common to all contigs databases. if a functional annotation source is not in this list does not mean 'a contigs database' is missing it, it means 'not all contigs databases' has it.